src/core/shards/serialization.cpp

```diff
@@ -296,69 +297,109 @@ struct LoadImage : public FileBase {
-      }
+      // image already given in bytes form
+      bytesValue = input.payload.bytesValue;
+      bytesSize = input.payload.bytesSize;
+    }
 
-      if (!_output.payload.imageValue.data) {
-        throw ActivationError(fmt::format("Failed to load image file {}", filename));
-      }
+    // check if it's a png (Using same signature check as stbi__check_png_header)
+    static const uint8_t png_sig[8] = { 137,80,78,71,13,10,26,10 };
+    is_png = (std::memcmp(bytesValue, png_sig, 8) == 0);
+
+    _output.valueType = SHType::Image;
+    int x, y, n;
+    switch(_bpp) {
+      case BPP::u8:
+        _output.payload.imageValue.data = 
+          reinterpret_cast<uint8_t *>(stbi_load_from_memory(bytesValue, static_cast<int>(bytesSize), &x, &y, &n, 0));
+        break;
+      case BPP::u16:
+        _output.payload.imageValue.data = 
+          reinterpret_cast<uint8_t *>(stbi_load_16_from_memory(bytesValue, static_cast<int>(bytesSize), &x, &y, &n, 0));
+        break;
+      default:
+        _output.payload.imageValue.data = 
+          reinterpret_cast<uint8_t *>(stbi_loadf_from_memory(bytesValue, static_cast<int>(bytesSize), &x, &y, &n, 0));
+        break;
+    }
+
+    if (!_output.payload.imageValue.data) {
+      throw ActivationError("Failed to load image file");
     }
+    
     _output.payload.imageValue.width = uint16_t(x);
     _output.payload.imageValue.height = uint16_t(y);
     _output.payload.imageValue.channels = uint16_t(n);
     switch (_bpp) {
-    case BPP::u16:
-      _output.payload.imageValue.flags = SHIMAGE_FLAGS_16BITS_INT;
-      break;
-    case BPP::f32:
-      _output.payload.imageValue.flags = SHIMAGE_FLAGS_32BITS_FLOAT;
-      break;
-    default:
-      _output.payload.imageValue.flags = 0;
-      break;
+      case BPP::u16:
+        _output.payload.imageValue.flags = SHIMAGE_FLAGS_16BITS_INT;
+        break;
+      case BPP::f32:
+        _output.payload.imageValue.flags = SHIMAGE_FLAGS_32BITS_FLOAT;
+        break;
+      default:
+        _output.payload.imageValue.flags = 0;
+        break;
+    }
+
+    // Premultiply the alpha channel if it is a PNG
+    auto pixsize = getPixelSize(_output);
+    if (is_png) {
+      // premultiply the alpha channel
+      switch(pixsize) {
+        case 1:
+          Imaging::premultiplyAlpha<uint8_t>(_output, _output, _output.payload.imageValue.width, _output.payload.imageValue.height);
+          break;
+        case 2:
+          Imaging::premultiplyAlpha<uint16_t>(_output, _output, _output.payload.imageValue.width, _output.payload.imageValue.height);
+          break;
+        case 4:
+          Imaging::premultiplyAlpha<float>(_output, _output, _output.payload.imageValue.width, _output.payload.imageValue.height);
+          break;
+      }
+    } else {
+      // if not a PNG, assume that it has already been premultiplied
+      _output.payload.imageValue.flags |= SHIMAGE_FLAGS_PREMULTIPLIED_ALPHA;
     }
```

Comment:
This is kind of a blanket solution though, I haven't fully verified that all of the image types supported by STBI, other than PNG, are all using premultiplied alpha channels. By attaching this flag to every image loaded by `LoadImage`, it also does mean that the `DemultiplyAlpha` shard will work on all images loaded through `LoadImage`, even if they aren't actually using premultiplied alpha channels. (The two new shards both do nothing if the SHIMAGE_FLAGS_PREMULTIPLIED_ALPHA is present for the `PremultiplyAlpha` shard and if it is not present for `DemultiplyAlpha`).

I think verifying that an image has actually been premultiplied or not may be impossible in [some cases](https://stackoverflow.com/q/9147973), so we can only do this based on specifications of the supported image file types

Reply:
I actually am thinking that we should remove this implicit premultiply. 
Always prefer explicit behaviours and it just takes an extra shard after LoadImage after all. It often will also be the case alpha is not even needed and such after all.
LoadImage should just load image basically simply.

Reply:
Right, not all images will even have an alpha channel. That makes sense.

But also to clarify, you mean removing the premultiply for PNGs as well right? I was thinking that some users might not know much about graphics and that PNGs specifically do not use premultiplied alpha channels and thus need to be premultiplied before it will render properly, so it might be easier to use if `LoadImage` handled it automatically. 

If we remove the implicit premultiply entirely, then shall we just write down the necessary additional step into the `LoadImage` documentation regarding PNGs and non-premultiplied alpha channel images as well as the tutorials?

Reply:
Yeah I think we should remove the implicit premultiply. It's easy to explain, and and it's easy to fix just need to add a shard after LoadImage after all.

Reply:
that or add a parameter to LoadImage, `Premultiply` or something so that it can be controlled

Reply:
`PremultiplyAlpha` parameter has been added in [5f6ec7d](https://github.com/fragcolor-xyz/shards/pull/682/commits/5f6ec7d04ef5ffe2924c72209e717ba3cd3f9d16) with default value of false